server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+              currentBlockId, readBufferSize, data);
+          offset = listReadInfo.offset;
+          // from next flush/cache buffer, read from 1st block
+          currentBlockId = Constants.INVALID_BLOCK_ID;
+          if (offset >= readBufferSize) {
+            break;
+          }
+        }
+      }
+    }
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      processCachedBlocks(offset, blocks, currentBlockId, readBufferSize, data);
+    }
+  }
+
+  private List<Long> sortFlushingEventId() {
+    List<Long> eventIdList = Lists.newArrayList(inFlushBlockMap.keySet());
+    eventIdList.sort((id1, id2) -> {
+      if (id1 > id2) {
+        return 1;
+      }
+      return -1;
+    });
+    return eventIdList;
+  }
+
+  private ListReadInfo processCachedBlocks(
+      int offset,
+      List<ShufflePartitionedBlock> cachedBlocks,
+      long lastBlockId,
+      long readBufferSize,
+      Object parameter) throws Exception {
+    int currentOffset = offset;
+    Callable<Integer> callable = null;
+    if (lastBlockId == Constants.INVALID_BLOCK_ID) {
+      // read from first block
+      for (ShufflePartitionedBlock block : cachedBlocks) {
+        // do action
+        if (parameter instanceof byte[]) {
+          callable = new CopyShuffleDataCall((byte[]) parameter, block, currentOffset);
+        } else if (parameter instanceof List) {
+          callable = new UpdateSegmentCall((List<BufferSegment>) parameter, block, currentOffset);
+        }
+        // update offset
+        currentOffset = callable.call();
+        if (currentOffset >= readBufferSize) {
+          break;
+        }
+      }
+      return new ListReadInfo(currentOffset, false);
+    } else {
+      // find lastBlockId, then read from next block
+      boolean foundBlockId = false;
+      for (ShufflePartitionedBlock block : cachedBlocks) {
+        if (!foundBlockId) {
+          if (block.getBlockId() == lastBlockId) {
+            foundBlockId = true;
+          }
+          continue;
+        }
+        if (parameter instanceof byte[]) {
+          callable = new CopyShuffleDataCall((byte[]) parameter, block, currentOffset);
+        } else if (parameter instanceof List) {
+          callable = new UpdateSegmentCall((List<BufferSegment>) parameter, block, currentOffset);
+        }
+        currentOffset = callable.call();
+        if (currentOffset >= readBufferSize) {
+          break;
+        }
+      }
+      return new ListReadInfo(currentOffset, foundBlockId);
+    }
+  }
+
+  class CopyShuffleDataCall implements Callable<Integer> {
+    final ShufflePartitionedBlock block;
+    final byte[] data;
+    final int offset;
+
+    CopyShuffleDataCall(byte[] data, ShufflePartitionedBlock block, int offset) {
+      this.block = block;
+      this.data = data;
+      this.offset = offset;
+    }
+
+    @Override
+    public Integer call() throws Exception {
+      try {
+        System.arraycopy(block.getData(), 0, data, offset, block.getLength());
```

Comment:
copy is necessary?

Reply:
yes, it's the transform like  `List<byte[]> -> byte[]`